### PR TITLE
(PUP-7042) Mark strings in parser

### PR DESCRIPTION
**Pull request by Magisus**
This commit marks user-facing error and info strings in
`lib/puppet/parser/*` for translation.

### CODE REVIEW

**Comment by Magisus** [comment]
Closing and reopening to fix CI.

**Comment by Magisus** [comment]
Closing and reopening to fix CI.

**Comment by McdonaldSeanp** [comment]
👍

File: lib/puppet/parser/compiler.rb [hunk 1]
```diff
@@ -91,7 +91,7 @@ def add_resource(scope, resource)
     @catalog.add_resource(resource)
 
     if not resource.class? and resource[:stage]
-      raise ArgumentError, "Only classes can set 'stage'; normal resources like #{resource} cannot change run stage"
+      raise ArgumentError, _("Only classes can set 'stage'; normal resources like %{resource} cannot change run stage") % { resource: resource }
```
Comment (MosesMendoza) [inline lib/puppet/parser/compiler.rb]: can you add a comment that `stage` is a key word in puppet?

File: lib/puppet/parser/compiler.rb [hunk 2]
```diff
@@ -79,7 +79,7 @@ def add_resource(scope, resource)
     if @current_app
       # We are in the process of pulling application components out that
       # apply to this node
-      Puppet.notice "Check #{resource}"
+      Puppet.notice _("Check %{resource}") % { resource: resource }
```
Comment (MosesMendoza) [inline lib/puppet/parser/compiler.rb]: I'm not sure how valuable this line and 39 above are for externalization... thoughts?

File: lib/puppet/parser/compiler.rb [hunk 3]
```diff
@@ -117,7 +117,7 @@ def assert_app_in_site(scope, resource)
           return if upstream.keys.map(&:type).include?('Site')
         end
       end
-      raise ArgumentError, "Application instances like '#{resource}' can only be contained within a Site"
+      raise ArgumentError, _("Application instances like '%{resource}' can only be contained within a Site") % { resource: resource }
```
Comment (MosesMendoza) [inline lib/puppet/parser/compiler.rb]: Can you add a note that `Site` is a key word in puppet and shouldn't be translated?

File: lib/puppet/parser/functions/generate.rb [hunk 4]
```diff
@@ -11,7 +11,7 @@
     generators, so all shell metacharacters are passed directly to
     the generator.") do |args|
 
-      raise Puppet::ParseError, "Generators must be fully qualified" unless Puppet::Util.absolute_path?(args[0])
+      raise Puppet::ParseError, _("Generators must be fully qualified") unless Puppet::Util.absolute_path?(args[0])
```
Comment (MosesMendoza) [inline lib/puppet/parser/functions/generate.rb]: Could you add a note that `fully qualified` here refers to a fully qualified file system path?

File: lib/puppet/parser/functions/include.rb [hunk 5]
```diff
@@ -28,6 +28,6 @@
 - Since 4.0.0 support for class and resource type values, absolute names
 - Since 4.7.0 returns an Array[Type[Class]] of all included classes 
 ") do |classes|
-  Puppet.warn_once(:deprecation, '3xfunction#include', "Calling function_include via the Scope class is deprecated. Use Scope#call_function instead")
   call_function('include', classes)
+  Puppet.warn_once(:deprecation, '3xfunction#include', _("Calling function_include via the Scope class is deprecated. Use Scope#call_function instead"))
```
Comment (MosesMendoza) [inline lib/puppet/parser/functions/include.rb]: Can you add a note that `function_include`, `Scope` and `Scope#call_function` reference function names in puppet?

File: lib/puppet/parser/compiler.rb (hunk 3 above)
Comment (Magisus) [inline lib/puppet/parser/compiler.rb]: yeah I agree, probably not super useful

File: lib/puppet/parser/compiler.rb [hunk 6]
```diff
@@ -485,7 +486,7 @@ def evaluate_ast_node
     end
 
     unless (astnode ||= krt.node("default"))
-      raise Puppet::ParseError, "Could not find node statement with name 'default' or '#{node.names.join(", ")}'"
+      raise Puppet::ParseError, _("Could not find node statement with name 'default' or '%{names}'") % { names: node.names.join(", ") }
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/compiler.rb]: Do we need to translate `'default'` ? Will users in different languages be using the word 'default' in their own language?

File: lib/puppet/parser/compiler.rb [hunk 7]
```diff
@@ -632,7 +633,7 @@ def finish
 
   def add_resource_metaparams
     unless main = catalog.resource(:class, :main)
-      raise "Couldn't find main"
+      raise _("Couldn't find main")
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/compiler.rb]: main is a keyword, does it need to stay in english?

File: lib/puppet/parser/functions.rb [hunk 8]
```diff
@@ -272,7 +272,7 @@ def get_function(name, environment)
 
   class Error
     def self.is4x(name)
-      raise Puppet::ParseError, "#{name}() can only be called using the 4.x function API. See Scope#call_function"
+      raise Puppet::ParseError, _("%{name}() can only be called using the 4.x function API. See Scope#call_function") % { name: name }
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/functions.rb]: `Scope#call_function` are keywords, should they be left in english?

File: lib/puppet/parser/functions/contain.rb [hunk 9]
```diff
@@ -25,6 +25,6 @@
 "
 ) do |classes|
   # Call the 4.x version of this function in case 3.x ruby code uses this function
-  Puppet.warn_once(:deprecation, '3xfunction#contain', "Calling function_contain via the Scope class is deprecated. Use Scope#call_function instead")
+  Puppet.warn_once(:deprecation, '3xfunction#contain', _("Calling function_contain via the Scope class is deprecated. Use Scope#call_function instead"))
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/functions/contain.rb]: `function_contain` `Scope` and `Scope#call_function` are all keywords, should they be left out of translation?

File: lib/puppet/parser/functions/create_resources.rb [hunk 10]
```diff
@@ -47,10 +47,10 @@
     final value of a parameter (just as when setting a parameter to `undef` in a puppet language
     resource declaration).
   ENDHEREDOC
-  raise ArgumentError, ("create_resources(): wrong number of arguments (#{args.length}; must be 2 or 3)") if args.length > 3
-  raise ArgumentError, ('create_resources(): second argument must be a hash') unless args[1].is_a?(Hash)
+  raise ArgumentError, (_("create_resources(): wrong number of arguments (%{count}; must be 2 or 3)") % { count: args.length }) if args.length > 3
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/functions/create_resources.rb]: `create_resources` is a function, should it be left out of translation?

File: lib/puppet/parser/functions/create_resources.rb [hunk 11]
```diff
@@ -47,10 +47,10 @@
...
+  raise ArgumentError, (_('create_resources(): second argument must be a hash')) unless args[1].is_a?(Hash)
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/functions/create_resources.rb]: `create_resources` is a function, should it be left out of translation?

File: lib/puppet/parser/functions/create_resources.rb [hunk 12]
```diff
@@ -47,10 +47,10 @@
...
   if args.length == 3
-    raise ArgumentError, ('create_resources(): third argument, if provided, must be a hash') unless args[2].is_a?(Hash)
+    raise ArgumentError, (_('create_resources(): third argument, if provided, must be a hash')) unless args[2].is_a?(Hash)
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/functions/create_resources.rb]: `create_resources` is a function, should it be left out of translation?

File: lib/puppet/parser/functions/include.rb [hunk 13]
```diff
@@ -28,6 +28,7 @@
 - Since 4.0.0 support for class and resource type values, absolute names
 - Since 4.7.0 returns an Array[Type[Class]] of all included classes 
 ") do |classes|
-  Puppet.warn_once(:deprecation, '3xfunction#include', "Calling function_include via the Scope class is deprecated. Use Scope#call_function instead")
   call_function('include', classes)
+  #TRANSLATORS "function_include", "Scope", and "Scope#call_function" refer to Puppet internals and should not be translated
+  Puppet.warn_once(:deprecation, '3xfunction#include', _("Calling function_include via the Scope class is deprecated. Use Scope#call_function instead"))
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/functions/include.rb]: `function_include` `Scope` and `Scope#call_function` are all keywords, should they be left out of translation?

File: lib/puppet/parser/functions/require.rb [hunk 14]
```diff
@@ -35,6 +35,6 @@ class otherstuff {
 - Since 4.0.0 Class and Resource types, absolute names
 - Since 4.7.0 Returns an Array[Type[Class]] with references to the required classes
 ") do |classes|
-  Puppet.warn_once(:deprecation, '3xfunction#require', "Calling function_require via the Scope class is deprecated. Use Scope#call_function instead")
   call_function('require', classes)
+  Puppet.warn_once(:deprecation, '3xfunction#require', _("Calling function_require via the Scope class is deprecated. Use Scope#call_function instead"))
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/functions/require.rb]: `function_require` `Scope` and `Scope#call_function` are all keywords, should they be left out of translation?

File: lib/puppet/parser/relationship.rb [hunk 15]
```diff
@@ -27,23 +27,23 @@ def initialize(source, target, type)
...
+    PARAM_MAP[type] || raise(ArgumentError, _("Invalid relationship type %{relationship_type}") % { relationship_type: type })
   end
 
   def mk_relationship(source, target, catalog)
     # There was once an assumption that this could be an array. These raise
     # assertions are here as a sanity check for 4.0 and can be removed after
     # a release or two
-    raise ArgumentError, "source shouldn't be an array" if source.is_a?(Array)
-    raise ArgumentError, "target shouldn't be an array" if target.is_a?(Array)
+    raise ArgumentError, _("source shouldn't be an array") if source.is_a?(Array)
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/relationship.rb]: source is a keyword, should it be left out of translation?

File: lib/puppet/parser/relationship.rb [hunk 16]
```diff
@@ -27,23 +27,23 @@ def initialize(source, target, type)
...
+    raise ArgumentError, _("target shouldn't be an array") if target.is_a?(Array)
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/relationship.rb]: target is a keyword, should it be left out of translation?

File: lib/puppet/parser/scope.rb [hunk 17]
```diff
@@ -883,7 +883,7 @@ def to_s
   # @deprecated use #pop_epehemeral
   # @api private
   def unset_ephemeral_var(level=:all)
-    Puppet.deprecation_warning('Method Parser::Scope#unset_ephemeral_var() is deprecated')
+    Puppet.deprecation_warning(_('Method Parser::Scope#unset_ephemeral_var() is deprecated'))
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/scope.rb]: `Parser::Scope#unset_ephemeral_var()` contains several function/classname keywords, should we comment about not translating it?

File: lib/puppet/parser/scope.rb [hunk 18]
```diff
@@ -1094,7 +1094,7 @@ def transform_and_assert_classnames(names)
         name.title.sub(/^([^:]{1,2})/, '::\1')
 
       when Puppet::Pops::Types::PHostClassType
-        raise ArgumentError, "Cannot use an unspecific Class[] Type" unless name.class_name
+        raise ArgumentError, _("Cannot use an unspecific Class[] Type") unless name.class_name
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/scope.rb]: `Class` is a keyword, should it be left in english?

File: lib/puppet/parser/scope.rb [hunk 19]
```diff
@@ -1120,13 +1120,13 @@ def call_function(func_name, args, &block)
 
   def assert_class_and_title(type_name, title)
     if type_name.nil? || type_name == ''
-      raise ArgumentError, "Cannot use an unspecific Resource[] where a Resource['class', name] is expected"
+      raise ArgumentError, _("Cannot use an unspecific Resource[] where a Resource['class', name] is expected")
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/scope.rb]: `Resource` is a keyword, should it be left in english?

File: lib/puppet/parser/scope.rb [hunk 20]
```diff
@@ -1120,13 +1120,13 @@ def call_function(func_name, args, &block)
...
     end
     unless type_name =~ /^[Cc]lass$/
-      raise ArgumentError, "Cannot use a Resource[#{type_name}] where a Resource['class', name] is expected"
+      raise ArgumentError, _("Cannot use a Resource[%{type_name}] where a Resource['class', name] is expected") % { type_name: type_name }
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/scope.rb]: Resource is a keyword, should it be left in english?

File: lib/puppet/parser/scope.rb [hunk 21]
```diff
@@ -1120,13 +1120,13 @@ def call_function(func_name, args, &block)
...
     end
     if title.nil?
-      raise ArgumentError, "Cannot use an unspecific Resource['class'] where a Resource['class', name] is expected"
+      raise ArgumentError, _("Cannot use an unspecific Resource['class'] where a Resource['class', name] is expected")
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/scope.rb]: Resource is a keyword, should it be left in english?

File: lib/puppet/parser/ast/resource.rb [hunk 22]
```diff
@@ -6,7 +6,7 @@ class Puppet::Parser::AST::Resource < Puppet::Parser::AST::Branch
   attr_accessor :type, :instances, :exported, :virtual
 
   def initialize(argshash)
-    Puppet.warn_once('deprecation', 'AST::Resource', 'Use of Puppet::Parser::AST::Resource is deprecated and not fully functional')
+    Puppet.warn_once('deprecation', 'AST::Resource', _('Use of Puppet::Parser::AST::Resource is deprecated and not fully functional'))
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/ast/resource.rb]: `Puppet::Parser::AST::Resource` contains several function/classname keywords, should we comment about not translating it?

File: lib/puppet/parser/ast/resource_instance.rb [hunk 23]
```diff
@@ -4,7 +4,7 @@
 class Puppet::Parser::AST::ResourceInstance < Puppet::Parser::AST::Branch
   attr_accessor :title, :parameters
   def initialize(argshash)
-    Puppet.warn_once('deprecation', 'AST::ResourceInstance', 'Use of Puppet::Parser::AST::ResourceInstance is deprecated')
+    Puppet.warn_once('deprecation', 'AST::ResourceInstance', _('Use of Puppet::Parser::AST::ResourceInstance is deprecated'))
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/ast/resource_instance.rb]: `Puppet::Parser::AST::ResourceInstance` contains several function/classname keywords, should we comment about not translating it?

File: lib/puppet/parser/ast/resourceparam.rb [hunk 24]
```diff
@@ -4,7 +4,7 @@ class Puppet::Parser::AST::ResourceParam < Puppet::Parser::AST::Branch
   attr_accessor :value, :param, :add
 
   def initialize(argshash)
-    Puppet.warn_once('deprecation', 'AST::ResourceParam', 'Use of Puppet::Parser::AST::ResourceParam is deprecated and not fully functional')
+    Puppet.warn_once('deprecation', 'AST::ResourceParam', _('Use of Puppet::Parser::AST::ResourceParam is deprecated and not fully functional'))
```
Comment (McdonaldSeanp) [inline lib/puppet/parser/ast/resourceparam.rb]: `Puppet::Parser::AST::ResourceParam` contains several function/classname keywords, should we comment about not translating it?

File: lib/puppet/parser/compiler.rb (hunk 6 above)
Comment (Magisus) [inline lib/puppet/parser/compiler.rb]: I think "default" is supplied by Puppet. Notice how it's passed as a hard-coded string in the line above? We're not translating things like that.

File: lib/puppet/parser/functions.rb (hunk 8 above)
Comment (Magisus) [inline lib/puppet/parser/functions.rb]: Can we just tell the translators to not translate anything that includes an `_`?

File: lib/puppet/parser/compiler.rb (hunk 6 above)
Comment (McdonaldSeanp) [inline lib/puppet/parser/compiler.rb]: Okay, that works

File: lib/puppet/parser/functions.rb (hunk 8 above)
Comment (McdonaldSeanp) [inline lib/puppet/parser/functions.rb]: That would work.

**Comment by Magisus** [comment]
Some of those suggestions seem very obviously to be things that shouldn't be translated. I want to have some faith in the translators. I updated the more ambiguous ones. Let me know if you guys think this approach is misguided.

File: lib/puppet/parser/functions.rb (hunk 8 above)
Comment (Magisus) [inline lib/puppet/parser/functions.rb]: Yeah. I sent an email to Larissa and Moses, we'll see what they have to say.

**Comment by McdonaldSeanp** [comment]
@Magisus that approach seems appropriate to me

**Comment by McdonaldSeanp** [comment]
👍

**Pull request merged**